plutus-emulator/src/Wallet/Emulator/NodeClient.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TupleSections         #-}
+{-# LANGUAGE TypeApplications      #-}
+module Wallet.Emulator.NodeClient where
+
+import           Conduit
+import           Control.Lens                 hiding (index)
+import           Control.Monad.Except
+import           Control.Monad.State          
+import           Data.Aeson                   (FromJSON, ToJSON)
+import           Data.List                    (partition)
+import           Data.Maybe                   (isNothing)
+import           Data.Traversable             (for)
+import           GHC.Generics                 (Generic)
+
+import           Ledger                       (Block, Blockchain, Slot (..), Tx (..), TxId, hashTx)
+import qualified Ledger.Index                 as Index
+import qualified Ledger.Interval              as Interval
+
+
+-- | Events produced by the blockchain emulator.
+data NodeClientEvent =
+    TxnSubmit TxId
+    -- ^ A transaction has been added to the pool of pending transactions.
+    | TxnValidate TxId
+    -- ^ A transaction has been validated and added to the blockchain.
+    | TxnValidationFail TxId Index.ValidationError
+    -- ^ A transaction failed  to validate.
+    | SlotAdd Slot
+    deriving (Eq, Show, Generic)
+
+instance FromJSON NodeClientEvent
+instance ToJSON NodeClientEvent
+
+-- | The result of validating a block.
+data ValidatedBlock = ValidatedBlock
+    { vlbValid  :: [Tx]
+    -- ^ The transactions that have been validated in this block.
+    , vlbEvents :: [NodeClientEvent]
+    -- ^ Transaction validation events for the transactions in this block.
+    , vlbRest   :: [Tx]
+    -- ^ The transactions that haven't been validated because the current slot is
+    --   not in their validation interval.
+    }
+
+-- | A pool of transactions which have yet to be validated.
+type TxPool = [Tx]
+
+data NodeClientState = NodeClientState {
```

Comment:
Let's call this the chain state. The other parts really are about the node client, but this represents the global state of the chain.

That also perhaps makes it clearer what's going on if we write something like:

`instance MonadState ChainState m => NodeClientEmulator m` or something: if you have direct access to the global chain state, then you can definitely emulate the node client behaviour.